activerecord/lib/active_record/connection_adapters/abstract_adapter.rb

```diff
@@ -797,6 +835,130 @@ def reconnect_can_restore_state?
           transaction_manager.restorable? && !@raw_connection_dirty
         end
 
+        # Lock the monitor, ensure we're properly connected and
+        # transactions are materialized, and then yield the underlying
+        # raw connection object.
+        #
+        # If +allow_retry+ is true, a connection-related exception will
+        # cause an automatic reconnect and re-run of the block, up to
+        # the connection's configured +connection_retries+ setting.
+        #
+        # If +uses_transaction+ is false, the block will be run without
+        # ensuring virtual transactions have been materialized in the DB
+        # server's state. The active transaction will also remain clean
+        # (if it is not already dirty), meaning it's able to be restored
+        # by reconnecting and opening an equivalent-depth set of new
+        # transactions. This should only be used by transaction control
+        # methods, and internal transaction-agnostic queries.
+        #
+        ###
+        #
+        # It's not the primary use case, so not something to optimize
+        # for, but note that this method does need to be re-entrant:
+        # +materialize_transactions+ will re-enter if it has work to do,
+        # and the yield block can also do so under some circumstances.
+        #
+        # In the latter case, we really ought to guarantee the inner
+        # call will not reconnect (which would interfere with the
+        # still-yielded connection in the outer block), but we currently
+        # provide no special enforcement there.
```

Comment:
Yeah, it's an (internal, implementation-scoped) comment noting a limitation of the current implementation. My intention was basically to acknowledge there's an obscure corner case that we're not handling; if a future reader is investigating an issue of that shape [because it's more likely than I expected, or because of other implementation changes], it's a helpful breadcrumb that they're on the right trail. I'll try to rephrase.